src/Servers/Kestrel/tools/CodeGenerator/KnownHeaders.cs

```diff
@@ -564,47 +572,76 @@ protected override bool CopyToFast(KeyValuePair<string, StringValues>[] array, i
+            ref readonly StringValues values = ref Unsafe.AsRef<StringValues>(null);
+
+            goto MoveNext;
 
-                    if({header.TestNotTempBit()})
+        OutputHeader:
+            {{
+                var valueCount = values.Count;
+                var headerKey = HeaderBytes.Slice(keyStart, keyLength);
+                for (var i = 0; i < valueCount; i++)
+                {{
+                    var value = values[i];
+                    if (value != null)
                     {{
-                        return;
+                        output.Write(headerKey);
+                        output.WriteAsciiNoValidation(value);
                     }}
-                    tempBits &= ~{1L << header.Index}L;
-                }}{(header.Identifier == "Server" ? $@"
-                if ((tempBits & {1L << 63}L) != 0)
+                }}
+
+                if (tempBits == 0)
                 {{
-                    output.Write(new ReadOnlySpan<byte>(_headerBytes, {loop.Headers.First(x => x.Identifier == "ContentLength").BytesOffset}, {loop.Headers.First(x => x.Identifier == "ContentLength").BytesCount}));
-                    output.WriteNumeric((ulong)ContentLength.Value);
+                    return;
+                }}
+            }}
 
-                    if((tempBits & ~{1L << 63}L) == 0)
-                    {{
-                        return;
-                    }}
-                    tempBits &= ~{1L << 63}L;
-                }}" : "")}")}
+        MoveNext:
+            // Compiled to Jump table
+            switch (next)
+            {{{Each(loop.Headers.OrderBy(h => !h.PrimaryHeader).Select((h, i) => (Header: h, Index: i)), hi => $@"
+                case {hi.Index}:
+                    goto Header{hi.Header.Identifier};")}
+                default:
+                    return;
+            }}
+            {Each(loop.Headers.OrderBy(h => !h.PrimaryHeader).Select((h, i) => (Header: h, Index: i)), hi => $@"
+        Header{hi.Header.Identifier}: // case {hi.Index}
+            if ({hi.Header.TestTempBit()})
+            {{
+                tempBits ^= {"0x" + (1L << hi.Header.Index).ToString("x")}L;{(hi.Header.Identifier != "ContentLength" ? $@"{(hi.Header.EnhancedSetter == false ? $@"
+                values = ref _headers._{hi.Header.Identifier};
+                keyStart = {hi.Header.BytesOffset};
+                keyLength = {hi.Header.BytesCount};
+                next = {hi.Index + 1};
+                goto OutputHeader;" : $@"
+                if (_headers._raw{hi.Header.Identifier} != null)
+                {{
+                    output.Write(_headers._raw{hi.Header.Identifier});
+                }}
+                else
+                {{
+                    values = ref _headers._{hi.Header.Identifier};
+                    keyStart = {hi.Header.BytesOffset};
+                    keyLength = {hi.Header.BytesCount};
+                    next = {hi.Index + 1};
+                    goto OutputHeader;
```

Comment:
Can we rewrite this using a loop instead of jumping to a new OutputHeader label. I'm fine with the gotos to switch cases, but this in particular really looks like it should just be a loop.

Reply:
Check out what ILspy rewrote it to: https://gist.github.com/davidfowl/60f990dd891d79e7e13549ff52870e59

Reply:
Make sense considering loops get compiled to branching operations which I assume are similar to the branching operations generated for gotos. How would ILSpy know this wasn't a loop to begin with?

The gist proves my point pretty well. The ILSpy decompiled code is easier to read (aside from some minor annoyances like putting the default case first) than the original generated source since you don't have to look for OutputHeader label or the associated gotos. Or as some might say, the "goto statement [is] considered harmful." :wink:

Reply:
Its full of gotos
```csharp
    goto case 1;
case 1:
```
🤣 

Anyway, just running tests on loop version

Reply:
Changed to loop; perf still looks good

Reply:
> Its full of gotos

I know... I almost didn't ask. At least the `goto case` is just a way to explicitly omit `break` inside a switch case. Similar to the reason we goto named labels outside of the switch statement.